src/MySqlConnector/MySqlClient/MySqlDataReader.cs

```diff
@@ -781,12 +830,32 @@ private enum State
 			NoMoreData,
 		}
 
+	    public async Task WaitDoneReadingAsync(CancellationToken cancellationToken=default(CancellationToken))
+	    {
+	        cancellationToken.ThrowIfCancellationRequested();
+	        await m_done_reading.WaitAsync(cancellationToken).ConfigureAwait(false);
+	        m_done_reading.Release();
+	    }
+
 		static readonly Task<bool> s_falseTask = Task.FromResult(false);
 		static readonly Task<bool> s_trueTask = Task.FromResult(true);
 
 		MySqlCommand m_command;
 		MySqlSession m_session;
-		State m_state;
+	    SemaphoreSlim m_done_reading = new SemaphoreSlim(1);
+	    State _m_state;
+	    State m_state
+	    {
+	        get { return _m_state; }
+	        set
+	        {
+	            if (value == State.None || value == State.NoMoreData)
+	            {
+	                m_done_reading.Release();
```

Comment:
Overall, I think it will be more reliable to move all this logic to Pomelo.EF, e.g., as in the proposed implementation here: https://github.com/bgrainger/Pomelo.EntityFrameworkCore.MySql/commit/7354ecf6dbf211721104943f878c07ad53c52581